### PR TITLE
Fix bad Scope#parent caching

### LINKED ISSUE #6057: babel compiled code use wrong module name if the module names 'args'
try it link: 
http://babeljs.io/repl/#?babili=false&evaluate=false&lineWrap=true&presets=es2015%2Creact%2Cstage-0&targets=&browsers=&builtIns=false&debug=false&code_lz=JYWwDg9gTgLgBAJQKYEMDGMA0cDecDCE4EAdkifAL5wBmURcA5FKhowNwBQokscKUAOYBnWvRBMArjGAAbYQHpJUWQoEiOnTkgAeveABMkNFJNnw0slMNEBBMGDi6Y5A6MLEyFXJzhO90DAAKgCeYEhwALxMjFp-aESQXjAAIsAGALIQkhQAFACUPn5-MAAWwMIAdLr6oeFR_EJVgkgwuYwwYUiM2ADuwCQGEL2VshBoKDKklaUsNPlcfpRxcCyDSFAFRcUsMMokcCRmsux-vnDLlEA

### Input Code
see try it link upside

### Babel Configuration (.babelrc, package.json, cli command)
see try it link upside

### Expected Behavior
expected to call `_args2.default.get`

### Current Behavior
call `args.get`, which is `undefined`

### Your Environment
see try it link upside

### PR DESCRIPTION
**Pull request by jridgewell**
Now, we traverse the parent paths until we find a parent scope.

Fixes #6057.

### CODE REVIEW

File: packages/babel-traverse/src/path/index.js [hunk 1]
```diff
@@ -113,6 +113,7 @@ export default class NodePath {
     }
 
     return ourScope;
+    return this.isScope() ? new Scope(this) : scope;
```
Comment (hzoo) [inline packages/babel-traverse/src/path/index.js]: this isn't intentional right since there's 2 returns?

Comment (jridgewell) [reply packages/babel-traverse/src/path/index.js]: Ha, that's probably why the tests were failing.

File: packages/babel-traverse/src/scope/index.js [hunk 2]
```diff
@@ -212,6 +189,15 @@ export default class Scope {
 
   static contextVariables = ["arguments", "undefined", "Infinity", "NaN"];
 
+  get parent() {
+    const parent = this.path.findParent(p => p.isScope());
+    return parent && parent.scope;
+  }
+
+  get parentBlock() {
+    return path.parent;
```
Comment (hzoo) [inline packages/babel-traverse/src/scope/index.js]: path isn't defined, `this.path`?

File: packages/babel-traverse/src/scope/index.js [hunk 3]
```diff
@@ -181,20 +162,19 @@ export default class Scope {
...
-    const cached = getCache(path, parentScope, this);
-    if (cached) return cached;
+    scopeCache.set(node, this);
 
     this.uid = uid++;
-    this.parent = parentScope;
-    this.hub = path.hub;
 
-    this.parentBlock = path.parent;
-    this.block = path.node;
```
Comment (danez) [inline packages/babel-traverse/src/scope/index.js]: Scope is also exposed to Plugins? Then removing props is a breaking change I guess.
Should land it before 7 then and add to migration guide :)

**Comment by danez** [comment]
We should have a babel performance repl, where we could enter code which gets transformed a couple of times and prints the average time it needed. Then we could easily check for performance.

**Comment by hzoo** [comment]
> babel performance repl

would be a great idea, could reuse some of the stuff we have already, already up for iterating fast on those kinds of things. there's also esbench.com cc @developit

**Comment by jridgewell** [comment]
There might be some performance regressions if a block node is used in two places in the AST, but otherwise this'll be the same.

**Pull request merged**